File: java/bench_mark/src/main/java/org/apache/tsfile/BenchMark.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.tsfile;

import org.apache.tsfile.enums.TSDataType;
import org.apache.tsfile.exception.read.ReadProcessException;
import org.apache.tsfile.exception.write.NoMeasurementException;
import org.apache.tsfile.exception.write.NoTableException;
import org.apache.tsfile.exception.write.WriteProcessException;
import org.apache.tsfile.file.metadata.ColumnSchema;
import org.apache.tsfile.file.metadata.ColumnSchemaBuilder;
import org.apache.tsfile.file.metadata.TableSchema;
import org.apache.tsfile.fileSystem.FSFactoryProducer;
import org.apache.tsfile.read.query.dataset.ResultSet;
import org.apache.tsfile.read.v4.ITsFileReader;
import org.apache.tsfile.read.v4.TsFileReaderBuilder;
import org.apache.tsfile.write.record.Tablet;
import org.apache.tsfile.write.v4.ITsFileWriter;
import org.apache.tsfile.write.v4.TsFileWriterBuilder;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

public class BenchMark {
  private static final Logger LOGGER = LoggerFactory.getLogger(BenchMark.class);

  public static void main(String[] args)
      throws IOException, ReadProcessException, NoTableException, NoMeasurementException {
    BenchMarkConf.printConfig();
    MemoryMonitor monitor = new MemoryMonitor();
    String path = "/tmp/tsfile_table_write_bench_mark.tsfile";
    File f = FSFactoryProducer.getFSFactory().getFile(path);
    if (f.exists()) {
      Files.delete(f.toPath());
    }
    monitor.recordMemoryUsage();
    List<String> column_names = new ArrayList<>();
    List<TSDataType> column_types = new ArrayList<>();
    List<ColumnSchema> columnSchemas = new ArrayList<>();
    columnSchemas.add(
        new ColumnSchemaBuilder()
            .name("TAG1")
            .dataType(TSDataType.STRING)
            .category(Tablet.ColumnCategory.TAG)
            .build());

    columnSchemas.add(
        new ColumnSchemaBuilder()
            .name("TAG2")
            .dataType(TSDataType.STRING)
            .category(Tablet.ColumnCategory.TAG)
            .build());
    column_names.add("TAG1");
    column_names.add("TAG2");
    column_types.add(TSDataType.STRING);
    column_types.add(TSDataType.STRING);

    int fieldIndex = 2;
    for (int i = 0; i < BenchMarkConf.FIELD_TYPE_VECTOR.size(); i++) {
      int count = BenchMarkConf.FIELD_TYPE_VECTOR.get(i);
      TSDataType dataType = BenchMarkConf.getTsDataType(i);
      for (int j = 0; j < count; j++) {
        columnSchemas.add(
            new ColumnSchemaBuilder()
                .name("FIELD" + fieldIndex)
                .dataType(dataType)
                .category(Tablet.ColumnCategory.FIELD)
                .build());
        column_names.add("FIELD" + fieldIndex);
        column_types.add(dataType);
        fieldIndex++;
      }
    }
    monitor.recordMemoryUsage();
    long totalPrepareTimeNs = 0;
    long totalWriteTimeNs = 0;
    long start = System.nanoTime();
    TableSchema tableSchema = new TableSchema("TestTable", columnSchemas);
    monitor.recordMemoryUsage();
    try (ITsFileWriter writer =
        new TsFileWriterBuilder().file(f).tableSchema(tableSchema).build()) {
      monitor.recordMemoryUsage();
      long timestamp = 0;
      for (int table_ind = 0; table_ind < BenchMarkConf.TABLET_NUM; table_ind++) {
        long prepareStartTime = System.nanoTime();
        Tablet tablet =
            new Tablet(
                column_names,
                column_types,
                BenchMarkConf.TAG1_NUM * BenchMarkConf.TAG2_NUM * BenchMarkConf.TIMESTAMP_PER_TAG);
        int row_count = 0;
        for (int tag1_ind = 0; tag1_ind < BenchMarkConf.TAG1_NUM; tag1_ind++) {
          for (int tag2_ind = 0; tag2_ind < BenchMarkConf.TAG2_NUM; tag2_ind++) {
            for (int row = 0; row < BenchMarkConf.TIMESTAMP_PER_TAG; row++) {
              tablet.addTimestamp(row_count, timestamp + row);
              tablet.addValue(row_count, 0, "tag1_" + tag1_ind);
              tablet.addValue(row_count, 1, "tag2_" + tag2_ind);

              for (int i = 2; i < column_types.size(); i++) {
                switch (column_types.get(i)) {
                  case INT32:
                    tablet.addValue(row_count, i, (int) timestamp);
                    break;
                  case INT64:
                    tablet.addValue(row_count, i, timestamp);
                    break;
                  case FLOAT:
                    tablet.addValue(row_count, i, (float) (timestamp * 1.1));
                    break;
                  case DOUBLE:
                    tablet.addValue(row_count, i, (double) timestamp * 1.1);
                    break;
                  case BOOLEAN:
                    tablet.addValue(row_count, i, timestamp % 2 == 0);
                  default:
                    //
                }
              }
              row_count++;
            }
          }
        }
        monitor.recordMemoryUsage();
        long prepareEndTime = System.nanoTime();

        totalPrepareTimeNs += (prepareEndTime - prepareStartTime);
        long writeStartTime = System.nanoTime();
        writer.write(tablet);
        monitor.recordMemoryUsage();
        long writeEndTime = System.nanoTime();
        totalWriteTimeNs += (writeEndTime - writeStartTime);
        timestamp += BenchMarkConf.TIMESTAMP_PER_TAG;
      }
    } catch (WriteProcessException e) {
      LOGGER.error("meet error in TsFileWrite ", e);
    }
    monitor.recordMemoryUsage();
    long end = System.nanoTime();
    double totalPrepareTimeSec = totalPrepareTimeNs / 1_000_000_000.0;
    double totalWriteTimeSec = totalWriteTimeNs / 1_000_000_000.0;
    double totalTimeSec = (end - start) / 1_000_000_000.0;

    long size = f.length();

    monitor.close();
    System.out.println("====================");
    System.out.println("finish bench mark for java");
    System.out.printf("tsfile size is %d bytes ~ %dKB%n", size, size / 1024);

    System.out.printf("prepare data time is %.6f s%n", totalPrepareTimeSec);
    System.out.printf("writing data time is %.6f s%n", totalWriteTimeSec);

    long totalPoints =
        (long) BenchMarkConf.TABLET_NUM
            * BenchMarkConf.TAG1_NUM
            * BenchMarkConf.TAG2_NUM
            * BenchMarkConf.TIMESTAMP_PER_TAG
            * column_names.size();
    double writingSpeed = totalPoints / totalTimeSec;
    System.out.printf("writing speed is %d points/s%n", (long) writingSpeed);

    System.out.printf("total time is %.6f s%n", totalTimeSec);
    System.out.println("====================");

    Integer row = 0;
    long read_start = System.nanoTime();
    try (ITsFileReader reader = new TsFileReaderBuilder().file(f).build()) {
      ResultSet resultSet = reader.query("TestTable", column_names, Long.MIN_VALUE, Long.MAX_VALUE);
      while (resultSet.next()) {
        row++;
      }
    }
    System.out.println("row is " + row);
    long read_end = System.nanoTime();
    double totalReadTimeSec = (read_end - read_start) / 1_000_000_000.0;
    System.out.printf("read time is %.6f s%n", totalReadTimeSec);
    double readSpeed = row * column_names.size() / totalReadTimeSec;
    System.out.printf("read speed is %.6f points/s %n", readSpeed);
  }
}

File: python/bench_mark/bench_mark.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
#
import os
from time import perf_counter

from tqdm import tqdm
import psutil
import csv
from tsfile import TSDataType, ColumnCategory
from tsfile import TableSchema, ColumnSchema
from tsfile import Tablet
from tsfile import TsFileTableWriter, TsFileReader


bench_mark_conf = {
    "tablet_num": 1000,
    "tag1_num": 1,
    "tag2_num": 10,
    "timestamp_per_tag": 1000,
    "field_type_vector": [1, 1, 1, 1, 1],
}

type_list = [TSDataType.INT32, TSDataType.INT64, TSDataType.FLOAT, TSDataType.DOUBLE, TSDataType.BOOLEAN]


def print_config():
    data_types_name = ["INT64", "INT32", "FLOAT", "DOUBLE", "BOOLEAN"]
    print("=====================")
    print("TsFile benchmark For Python")
    print("Schema Configuration:")
    print(f"Tag Column num: {2}")
    print(f"TAG1 num: {bench_mark_conf['tag1_num']} TAG2 num: {bench_mark_conf['tag2_num']}\n")

    print("Filed Column and types: ")
    column_num = 0
    for i in range(5):
        print(f"{data_types_name[i]}x{bench_mark_conf['field_type_vector'][i]}  ", end="")
        column_num += bench_mark_conf['field_type_vector'][i]

    print("\n")
    print(f"Tablet num: {bench_mark_conf['tablet_num']}")
    print(f"Tablet row num per tag: {bench_mark_conf['timestamp_per_tag']}")

    total_points = (bench_mark_conf['tablet_num'] *
                    bench_mark_conf['tag1_num'] *
                    bench_mark_conf['tag2_num'] *
                    bench_mark_conf['timestamp_per_tag'] *
                    column_num)
    print(f"Total points is {total_points}")
    print("======")

column_name = []

def get_memory_usage_kb():
    process = psutil.Process(os.getpid())
    return process.memory_info().rss // 1024

def bench_mark_write():
    csv_file = "memory_usage_python.csv"

    csv_writer = csv.writer(open(csv_file, "w"))
    csv_writer.writerow(["iter_num", "memory_usage(kb)"])
    iter_num = 0
    print_config()
    column_schema_list = []
    column_datat_type = []
    column_schema_list.append(ColumnSchema("TAG1", TSDataType.STRING, ColumnCategory.TAG))
    column_name.append("TAG1")
    column_datat_type.append(TSDataType.STRING)
    column_schema_list.append(ColumnSchema("TAG2", TSDataType.STRING, ColumnCategory.TAG))
    column_name.append("TAG2")
    column_datat_type.append(TSDataType.STRING)


    i = 2
    for count, type in zip(bench_mark_conf["field_type_vector"], type_list):
        for _ in range(count):
            column_schema_list.append(ColumnSchema("FIELD" + str(i), type, ColumnCategory.FIELD))
            column_name.append("FIELD" + str(i))
            column_datat_type.append(type)
            i = i + 1

    timestamp = 0
    table_schema = TableSchema("TestTable", column_schema_list)
    start = perf_counter()
    prepare_time = 0
    writing_time = 0
    csv_writer.writerow([iter_num, get_memory_usage_kb()])
    iter_num += 1
    with TsFileTableWriter("tsfile_table_write_bench_mark.tsfile", table_schema) as writer:
        csv_writer.writerow([iter_num, get_memory_usage_kb()])
        iter_num += 1
        for i in tqdm(range(bench_mark_conf["tablet_num"]), desc="Tablets"):
            csv_writer.writerow([iter_num, get_memory_usage_kb()])
            iter_num += 1
            row_num = 0
            prepare_start = perf_counter()
            tablet = Tablet(column_name, column_datat_type,
                            bench_mark_conf["timestamp_per_tag"] * bench_mark_conf["tag1_num"] *
                            bench_mark_conf["tag2_num"])

            for j in range(bench_mark_conf["tag1_num"]):
                for k in range(bench_mark_conf["tag2_num"]):
                    for row in range(bench_mark_conf["timestamp_per_tag"]):
                        tablet.add_timestamp(row_num, timestamp + row)
                        tablet.add_value_by_index(0, row_num, "tag1_" + str(j))
                        tablet.add_value_by_index(1, row_num, "tag2_" + str(k))
                        for col in range(2, len(column_name)):
                            if column_datat_type[col] == TSDataType.INT32:
                                tablet.add_value_by_index(col, row_num, timestamp)
                            elif column_datat_type[col] == TSDataType.INT64:
                                tablet.add_value_by_index(col, row_num, timestamp)
                            elif column_datat_type[col] == TSDataType.FLOAT:
                                tablet.add_value_by_index(col, row_num, timestamp * 1.1)
                            elif column_datat_type[col] == TSDataType.DOUBLE:
                                tablet.add_value_by_index(col, row_num, timestamp * 1.1)
                            elif column_datat_type[col] == TSDataType.BOOLEAN:
                                tablet.add_value_by_index(col, row_num, timestamp % 2 == 0)
                        row_num = row_num + 1

            prepare_time += perf_counter() - prepare_start
            write_start = perf_counter()
            csv_writer.writerow([iter_num, get_memory_usage_kb()])
            iter_num += 1
            writer.write_table(tablet)
            writing_time += perf_counter() - write_start
            timestamp = timestamp + bench_mark_conf["timestamp_per_tag"]
            csv_writer.writerow([iter_num, get_memory_usage_kb()])
            iter_num += 1

    csv_writer.writerow([iter_num, get_memory_usage_kb()])
    iter_num += 1
    end = perf_counter()
    total_time = end - start
    size = os.path.getsize("tsfile_table_write_bench_mark.tsfile")

    total_points = bench_mark_conf["tablet_num"] * bench_mark_conf["tag1_num"] * bench_mark_conf["tag2_num"] * \
                   bench_mark_conf["timestamp_per_tag"] * len(column_name)

    print("finish bench mark for python")
    print(f"tsfile size is {size} bytes ~ {size // 1024}KB")

    print(f"prepare data time is {prepare_time:.6f} s")
    print(f"writing data time is {writing_time:.6f} s")
    print(f" total_time is {total_time} s")
    writing_speed = int(total_points / (prepare_time + writing_time))
    print(f"writing speed is {writing_speed} points/s")

    total_time_seconds = (end - start)
    print(f"total time is {total_time_seconds:.6f} s")

def bench_mark_read():
    start = perf_counter()
    row = 0
    with TsFileReader("tsfile_table_write_bench_mark.tsfile") as reader:
        result = reader.query_table("TestTable", column_name)
        first = True
        while result.next():
            row = row + 1

    end = perf_counter()
    total_time = end - start
    reading_speed = int(row * len(column_name) / total_time)
    print("total row is ", row)
    print(f"reading data time is {total_time} s")
    print(f"reading data speed is {reading_speed} points/s")


bench_mark_write()
bench_mark_read()
